kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/flow/RetryFlow.kt

```diff
@@ -0,0 +1,80 @@
+package io.provenance.kafka.coroutines.retry.flow
+
+import io.provenance.kafka.coroutines.retry.FlowRetry
+import io.provenance.kafka.coroutines.retry.RetryStrategy
+import io.provenance.kafka.coroutines.retry.defaultRetryStrategies
+import io.provenance.kafka.coroutines.retry.invert
+import io.provenance.kafka.coroutines.retry.store.RetryRecord
+import io.provenance.kafka.coroutines.retry.tryMap
+import java.time.OffsetDateTime
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.toJavaDuration
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onStart
+import mu.KotlinLogging
+
+val DEFAULT_RETRY_INTERVAL = 10.seconds
+
+/**
+ * Create a retrying [Flow].
+ *
+ * Using [FlowRetry.produceNext] as the data source, feed retry records into [FlowRetry.retry].
+ * * If successful, route through [FlowRetry.onSuccess].
+ * * If failed, route through [FlowRetry.onFailure].
+ *
+ * Once a record is successfully processed, emit the data element out to the flow.
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+fun <T> retryFlow(
+    flowRetry: FlowRetry<T>,
+    retryInterval: Duration = DEFAULT_RETRY_INTERVAL,
+    retryStrategies: List<RetryStrategy> = defaultRetryStrategies
+): Flow<T> {
+    val log = KotlinLogging.logger {}
+    val strategies = retryStrategies.invert()
+
+    return pollingFlow(retryInterval) {
+        for (strategy in strategies) {
+            val lastAttempted = OffsetDateTime.now().minus(strategy.value.lastAttempted.toJavaDuration())
```

Comment:
im dumb, how does retry strategy work?

Reply:
for example:
`RetryStrategy(5, 10.minutes), RetryStrategy(4, 1.days)` would retry 5 times, with 10 minutes in between; once those are exhausted, retry 4 times, with 1 day in between. it's a defined back-off strategy.